full-stack-tests/presentation/src/learning-snippets/Hierarchies.test.ts

```diff
@@ -0,0 +1,329 @@
+            type: "B",
+            label: "B",
+          }],
+        }],
+      };
+      // __PUBLISH_EXTRACT_END__
+
+      // Expect A root node with a B child
+      const rootNodes = await Presentation.presentation.getNodes({ imodel, rulesetOrId: ruleset });
+      expect(rootNodes).to.containSubset([{
+        label: { displayValue: "A" },
+      }]);
+
+      const childNodes = await Presentation.presentation.getNodes({ imodel, rulesetOrId: ruleset, parentKey: rootNodes[0].key });
+      expect(childNodes).to.containSubset([{
+        label: { displayValue: "B" },
+      }]);
+    });
+
+    it("uses ruleset variables in rule condition", async () => {
+      // __PUBLISH_EXTRACT_START__ Hierarchies.Condition.RulesetVariables.Ruleset
+      // The ruleset has two root node rules that return "A" and "B" nodes. Both rules can be enabled or
+      // disabled with a ruleset variable.
+      const ruleset: Ruleset = {
+        id: "example",
+        rules: [{
+          ruleType: RuleTypes.RootNodes,
+          condition: `GetVariableBoolValue("DISPLAY_A_NODES")`,
+          specifications: [{
+            specType: ChildNodeSpecificationTypes.CustomNode,
+            type: "A",
+            label: "A",
+          }],
+        }, {
+          ruleType: RuleTypes.RootNodes,
+          condition: `GetVariableBoolValue("DISPLAY_B_NODES")`,
+          specifications: [{
+            specType: ChildNodeSpecificationTypes.CustomNode,
+            type: "B",
+            label: "B",
+          }],
+        }],
+      };
+      // __PUBLISH_EXTRACT_END__
+
+      // No variables set - no nodes
+      let nodes = await Presentation.presentation.getNodes({ imodel, rulesetOrId: ruleset });
+      expect(nodes).to.be.empty;
+
+      // Set DISPLAY_B_NODES to get node B
+      await Presentation.presentation.vars(ruleset.id).setBool("DISPLAY_B_NODES", true);
+      nodes = await Presentation.presentation.getNodes({ imodel, rulesetOrId: ruleset });
+      expect(nodes.length).to.eq(1);
+      expect(nodes).to.containSubset([{
+        label: { displayValue: "B" },
+      }]);
+
+      // Set DISPLAY_A_NODES to also get node A
+      await Presentation.presentation.vars(ruleset.id).setBool("DISPLAY_A_NODES", true);
+      nodes = await Presentation.presentation.getNodes({ imodel, rulesetOrId: ruleset });
+      expect(nodes.length).to.eq(2);
+      expect(nodes).to.containSubset([{
+        label: { displayValue: "A" },
+      }, {
+        label: { displayValue: "B" },
+      }]);
+    });
+
+    it("uses `requiredSchemas` attribute", async () => {
+      // __PUBLISH_EXTRACT_START__ Hierarchies.RequiredSchemas.Ruleset
+      // The ruleset has one root node rule that returns `bis.ExternalSourceAspect` instances. The
+      // ECClass was introduced in BisCore version 1.0.2, so the rule needs a `requiredSchemas` attribute
+      // to only use the rule if the version meets the requirement.
+      const ruleset: Ruleset = {

```

Comment:
Good question. I personally check [here](https://github.com/iTwin/bis-schemas/tree/master/Domains/Core/Released) to find what schema version some ECClass appeared first. But that's definitely not very convenient...

@ColinKerr, is there any better way? For example an ECSchema specific changelog or another way to find which ECSchema version some ECClass/ECRelationship/ECProperty appeared first?

Reply:
The linked repo is not yet public but it will be public soon.  It is at the moment the best way to figure out when a concept was added to a schema.  Making an easier way to find out what was added in each version is a reasonable request and something I think we can generate programmatically.  I will add it to our backlog.  Thanks for the suggestion.

Reply:
@roluk sounds like we'll have to leave this as is until the repo becomes public. We can update the comment afterwards.